flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/adaptive/AdaptiveJoinOperatorGeneratorTest.java

```diff
@@ -0,0 +1,383 @@
+        assertThat(adaptiveJoin.enrichAndCheckBroadcast(10, 2, 5))
+                .isEqualTo(new Tuple2<>(false, true));
+        assertThat(adaptiveJoin.enrichAndCheckBroadcast(2, 10, 1))
+                .isEqualTo(new Tuple2<>(false, true));
+        assertThat(adaptiveJoin.enrichAndCheckBroadcast(10, 2, 1))
+                .isEqualTo(new Tuple2<>(false, true));
+    }
+
+    @Test
+    void testLeftJoinCheckBroadcast() {
+        testBuildRightCheckBroadcast(FlinkJoinType.LEFT);
+    }
+
+    @Test
+    void testSemiJoinCheckBroadcast() {
+        testBuildRightCheckBroadcast(FlinkJoinType.SEMI);
+    }
+
+    @Test
+    void testAntiJoinCheckBroadcast() {
+        testBuildRightCheckBroadcast(FlinkJoinType.ANTI);
+    }
+
+    @Test
+    void testFULLJoinCheckBroadcast() {
+        AdaptiveJoin adaptiveJoin =
+                genAdaptiveJoin(FlinkJoinType.FULL, OperatorType.ShuffleHashJoin);
+        assertThatThrownBy(() -> adaptiveJoin.enrichAndCheckBroadcast(2, 10, 5))
+                .hasMessageContaining("Unexpected join type");
+    }
+
+    void testBuildRightCheckBroadcast(FlinkJoinType joinType) {
+        AdaptiveJoin adaptiveJoin = genAdaptiveJoin(joinType, OperatorType.ShuffleHashJoin);
+        assertThat(adaptiveJoin.enrichAndCheckBroadcast(2, 10, 5))
+                .isEqualTo(new Tuple2<>(false, true));
+        assertThat(adaptiveJoin.enrichAndCheckBroadcast(10, 2, 5))
+                .isEqualTo(new Tuple2<>(true, false));
+        assertThat(adaptiveJoin.enrichAndCheckBroadcast(2, 10, 1))
+                .isEqualTo(new Tuple2<>(false, true));
+        assertThat(adaptiveJoin.enrichAndCheckBroadcast(10, 2, 1))
+                .isEqualTo(new Tuple2<>(false, false));
+
+        adaptiveJoin = genAdaptiveJoin(joinType, OperatorType.SortMergeJoin);
+        assertThat(adaptiveJoin.enrichAndCheckBroadcast(2, 10, 5))
+                .isEqualTo(new Tuple2<>(false, true));
+        assertThat(adaptiveJoin.enrichAndCheckBroadcast(10, 2, 5))
+                .isEqualTo(new Tuple2<>(true, false));
+        assertThat(adaptiveJoin.enrichAndCheckBroadcast(2, 10, 1))
+                .isEqualTo(new Tuple2<>(false, true));
+        assertThat(adaptiveJoin.enrichAndCheckBroadcast(10, 2, 1))
+                .isEqualTo(new Tuple2<>(false, true));
+    }
+
+    // ---------------------- Test the correctness of the generated join operator -----------------
+    @Test
+    void testInnerJoinCorrectness() throws Exception {
+        // Shuffle Hash join
+        testInnerJoin(true, OperatorType.ShuffleHashJoin, false);
+        testInnerJoin(false, OperatorType.ShuffleHashJoin, false);
+        // Sort merge join
+        testInnerJoin(true, OperatorType.SortMergeJoin, false);
+        testInnerJoin(true, OperatorType.SortMergeJoin, true);
+        // Broadcast Hash join
+        testInnerJoin(true, OperatorType.ShuffleHashJoin, true);
+        testInnerJoin(false, OperatorType.ShuffleHashJoin, true);
```

Comment:
Maybe we can re-organize the assertions to make it a bit more clear to understand what joins will be generated in different cases.
- For ShuffleHashJoin
    - all cases to ShuffleHashJoin
    - all cases to BroadcastHashJoin
- ForSortMergeJoin
    - all cases to SortMergeJoin
    - all cases to BroadcastHashJoin

Reply:
And maybe also verify the generated operator type?